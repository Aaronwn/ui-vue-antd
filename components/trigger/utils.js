function isPointsEq (a1, a2) {
  return a1[0] === a2[0] && a1[1] === a2[1]
}

export function getAlignFromPlacement (builtinPlacements, placementStr, align) {
  const baseAlign = builtinPlacements[placementStr] || {}
  return {
    ...baseAlign,
    ...align,
  }
}

export function getPopupClassNameFromAlign (builtinPlacements, prefixCls, align) {
  const points = align.points
  for (const placement in builtinPlacements) {
    if (builtinPlacements.hasOwnProperty(placement)) {
      if (isPointsEq(builtinPlacements[placement].points, points)) {
        return `${prefixCls}-placement-${placement}`
      }
    }
  }
  return ''
}
export function noop () {
}
