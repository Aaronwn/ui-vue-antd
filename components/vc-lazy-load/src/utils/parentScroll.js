const style = (element, prop) => {
  let styleVal = ''
  if (typeof getComputedStyle !== 'undefined') {
    styleVal = window.getComputedStyle(element, null).getPropertyValue(prop)
  } else {
    styleVal = element.style[prop]
  }
  return styleVal
}

const overflow = (element) =>
  style(element, 'overflow') + style(element, 'overflow-y') + style(element, 'overflow-x')

const scrollParent = (element) => {
  if (!(element instanceof window.HTMLElement)) {
    return window
  }

  let parent = element

  while (parent) {
    if (parent === document.body || parent === document.documentElement) {
      break
    }

    if (!parent.parentNode) {
      break
    }
    if (/(scroll|auto)/.test(overflow(parent))) {
      return parent
    }

    parent = parent.parentNode
  }

  return window
}

export default scrollParent
