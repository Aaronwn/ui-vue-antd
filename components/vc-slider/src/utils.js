import keyCode from '../../_util/KeyCode'

export function isEventFromHandle (e, handles) {
  console.log(Object.keys(handles)
    .some(key => e.target === handles[key].$el))
  return Object.keys(handles)
    .some(key => e.target === handles[key].$el)
}

export function isValueOutOfRange (value, { min, max }) {
  return value < min || value > max
}

export function isNotTouchEvent (e) {
  return e.touches.length > 1 ||
    (e.type.toLowerCase() === 'touchend' && e.touches.length > 0)
}

export function getClosestPoint (val, { marks, step, min }) {
  const points = Object.keys(marks).map(parseFloat)
  if (step !== null) {
    const closestStep =
            Math.round((val - min) / step) * step + min
    points.push(closestStep)
  }
  const diffs = points.map(point => Math.abs(val - point))
  return points[diffs.indexOf(Math.min(...diffs))]
}

export function getPrecision (step) {
  const stepString = step.toString()
  let precision = 0
  if (stepString.indexOf('.') >= 0) {
    precision = stepString.length - stepString.indexOf('.') - 1
  }
  return precision
}

export function getMousePosition (vertical, e) {
  return vertical ? e.clientY : e.pageX
}

export function getTouchPosition (vertical, e) {
  return vertical ? e.touches[0].clientY : e.touches[0].pageX
}

export function getHandleCenterPosition (vertical, handle) {
  const coords = handle.getBoundingClientRect()
  return vertical
    ? coords.top + (coords.height * 0.5)
    : coords.left + (coords.width * 0.5)
}

export function ensureValueInRange (val, { max, min }) {
  if (val <= min) {
    return min
  }
  if (val >= max) {
    return max
  }
  return val
}

export function ensureValuePrecision (val, props) {
  const { step } = props
  const closestPoint = getClosestPoint(val, props)
  return step === null ? closestPoint
    : parseFloat(closestPoint.toFixed(getPrecision(step)))
}

export function pauseEvent (e) {
  e.stopPropagation()
  e.preventDefault()
}

export function getKeyboardValueMutator (e) {
  switch (e.keyCode) {
    case keyCode.UP:
    case keyCode.RIGHT:
      return (value, props) => value + props.step

    case keyCode.DOWN:
    case keyCode.LEFT:
      return (value, props) => value - props.step

    case keyCode.END: return (value, props) => props.max
    case keyCode.HOME: return (value, props) => props.min
    case keyCode.PAGE_UP: return (value, props) => value + props.step * 2
    case keyCode.PAGE_DOWN: return (value, props) => value - props.step * 2

    default: return undefined
  }
}

export function getComponentProps (obj, prop) {
  if (obj[prop]) {
    return obj
  } else if (obj.$children.length) {
    const len = obj.$children.length
    for (let i = 0; i < len; i++) {
      if (obj.$children[i][prop]) {
        return obj.$children[i]
      } else if (obj.$children[i].$children.length) {
        return getComponentProps(obj.$children[i], prop)
      }
    }
  }
}
