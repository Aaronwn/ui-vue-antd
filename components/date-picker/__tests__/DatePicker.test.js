import { mount } from '@vue/test-utils'
import { asyncExpect } from '@/tests/utils'
import moment from 'moment'
import MockDate from 'mockdate'
import DatePicker from '..'
import {
  selectDateFromBody,
  openPanel,
  clearInput,
  nextYear,
  nextMonth,
  hasSelected,
  $$,
} from './utils'
import focusTest from '../../../tests/shared/focusTest'

describe('DatePicker', () => {
  focusTest(DatePicker)

  beforeEach(() => {
    document.body.outerHTML = ''
    MockDate.set(moment('2016-11-22'))
  })

  afterEach(() => {
    MockDate.reset()
  })

  it('prop locale should works', async () => {
    const locale = {
      lang: {
        placeholder: 'Избери дата',
        rangePlaceholder: [
          'Начална дата',
          'Крайна дата',
        ],
        today: 'Днес',
        now: 'Сега',
        backToToday: 'Към днес',
        ok: 'Добре',
        clear: 'Изчистване',
        month: 'Месец',
        year: 'Година',
        timeSelect: 'Избор на час',
        dateSelect: 'Избор на дата',
        monthSelect: 'Избор на месец',
        yearSelect: 'Избор на година',
        decadeSelect: 'Десетилетие',
        previousMonth: 'Предишен месец (PageUp)',
        nextMonth: 'Следващ месец (PageDown)',
        previousYear: 'Последна година (Control + left)',
        nextYear: 'Следваща година (Control + right)',
        previousDecade: 'Предишно десетилетие',
        nextDecade: 'Следващо десетилетие',
        previousCentury: 'Последен век',
        nextCentury: 'Следващ век',
        yearFormat: 'YYYY',
        dateFormat: 'D M YYYY',
        dayFormat: 'D',
        dateTimeFormat: 'D M YYYY HH:mm:ss',
        monthBeforeYear: true,
      },
      timePickerLocale: {
        placeholder: 'Избор на час',
      },
    }
    const birthday = moment('2000-01-01', 'YYYY-MM-DD')
    const wrapper = mount({
      render () {
        return <DatePicker open locale={locale} value={birthday} />
      },
    })
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  // Fix https://github.com/ant-design/ant-design/issues/8885
  it('control value after panel closed', async () => {
    const Test = {
      data () {
        return {
          cleared: false,
          value: moment(),
        }
      },
      methods: {
        onChange  (value) {
          let cleared = this.cleared

          if (cleared) {
            value = moment(moment(value).format('YYYY-MM-DD 12:12:12'))
            cleared = false
          }

          if (!value) {
            cleared = true
          }
          this.value = value
          this.cleared = cleared
        },
      },
      render () {
        return (
          <DatePicker
            showTime
            value={this.value}
            format='YYYY-MM-DD HH:mm:ss'
            onChange={this.onChange}
          />
        )
      },
    }

    const wrapper = mount(Test, { sync: false, attachToDocument: true })
    await asyncExpect(() => {
      // clear input
      clearInput(wrapper)
    })
    await asyncExpect(() => {
      openPanel(wrapper)
    })
    await asyncExpect(() => {
      selectDateFromBody(moment('2016-11-13'))
    }, 0)
    await asyncExpect(() => {
      expect($$('.ant-calendar-input')[0].value).toBe('2016-11-13 12:12:12')
    })
    await asyncExpect(() => {
      selectDateFromBody(moment('2016-11-14'))
    })
    await asyncExpect(() => {
      expect($$('.ant-calendar-input')[0].value).toBe('2016-11-14 12:12:12')
    })
    await asyncExpect(() => {

    })
  })

  it('triggers onChange only when date was selected', async () => {
    const handleChange = jest.fn()
    const wrapper = mount({
      render () {
        return <DatePicker onChange={handleChange} />
      },
    }, { sync: false, attachToDocument: true })
    await asyncExpect(() => {
      openPanel(wrapper)
    })
    await asyncExpect(() => {
      nextYear()
    }, 0)
    await asyncExpect(() => {
      expect(handleChange).not.toBeCalled()
    })
    await asyncExpect(() => {
      nextMonth()
    })
    await asyncExpect(() => {
      expect(handleChange).not.toBeCalled()
    })
    await asyncExpect(() => {
      selectDateFromBody(moment('2017-12-22'))
    })
    await asyncExpect(() => {
      expect(handleChange).toBeCalled()
    })
    await asyncExpect(() => {

    })
  })

  it('clear input', async () => {
    const wrapper = mount(DatePicker, { sync: false, attachToDocument: true })
    await asyncExpect(() => {
      openPanel(wrapper)
    })
    await asyncExpect(() => {
      selectDateFromBody(moment('2016-11-23'))
    }, 0)
    await asyncExpect(() => {
      clearInput(wrapper)
    }, 0)
    await asyncExpect(() => {
      openPanel(wrapper)
    }, 0)
    await asyncExpect(() => {
      expect(hasSelected(wrapper, moment('2016-11-22'))).toBe(true)
    }, 0)
  })
})
