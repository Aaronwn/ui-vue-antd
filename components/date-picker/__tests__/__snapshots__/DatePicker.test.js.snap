// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker prop locale should works 1`] = `
<span class="ant-calendar-picker"><div class=""><input readonly="true" placeholder="Избери дата" class="ant-calendar-picker-input ant-input"><i class="anticon anticon-cross-circle ant-calendar-picker-clear"></i><span class="ant-calendar-picker-icon"></span></div>
</span>
`;
